core/http/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/ActiveTransactionRegistry.java

```diff
@@ -216,12 +244,39 @@ public RepositoryConnection getTransactionConnection(UUID transactionId)
 	 *        a transaction identifier.
 	 */
 	public void returnTransactionConnection(UUID transactionId) {
+		final CacheEntry entry = primaryCache.getIfPresent(transactionId);
+		if (entry != null) {
+			bumpSecondaryCache(transactionId, entry);
+			final ReentrantLock txnLock = entry.getLock();
+			if (txnLock.isHeldByCurrentThread()) {
+				txnLock.unlock();
+			}
+		}
+	}
 
-		final CacheEntry entry = activeConnections.getIfPresent(transactionId);
+	/**
+	 * Checks if the given transaction entry is still in the secondary cache (resetting its last access time
+	 * in the process) and if not reinsert it.
+	 * 
+	 * @param transactionId
+	 *        the id for the transaction to check
+	 * @param entry
+	 *        the cache entry to insert if necessary.
+	 */
+	private void bumpSecondaryCache(UUID transactionId, final CacheEntry entry) {
```

Comment:
'bump' is _totally_ a [technical term](http://www.urbandictionary.com/define.php?term=bump) :)

Seriously though, you're right, I'll  change it to 'updateSecondaryCache' .